components/api_request_helper/api_request_helper_unittest.cc

```diff
@@ -0,0 +1,131 @@
+    const absl::optional<std::string>& converted_response,
+    const std::string& raw_response) {
+  EXPECT_EQ(expected_raw_response, raw_response);
+  return converted_response;
+}
+}  // namespace
+
+class ApiRequestHelperUnitTest : public testing::Test {
+ public:
+  ApiRequestHelperUnitTest()
+      : shared_url_loader_factory_(
+            base::MakeRefCounted<network::WeakWrapperSharedURLLoaderFactory>(
+                &url_loader_factory_)) {
+    api_request_helper_.reset(new APIRequestHelper(
+        net::NetworkTrafficAnnotationTag(TRAFFIC_ANNOTATION_FOR_TESTS),
+        shared_url_loader_factory_));
+  }
+  ~ApiRequestHelperUnitTest() override = default;
+
+  void SetInterceptor(const std::string& expected_method,
+                      const GURL& expected_url,
+                      const std::string& content_to_respond) {
+    url_loader_factory_.SetInterceptor(base::BindLambdaForTesting(
+        [&, expected_method, expected_url,
+         content_to_respond](const network::ResourceRequest& request) {
+          url_loader_factory_.ClearResponses();
+          EXPECT_EQ(request.url, expected_url);
+          EXPECT_EQ(request.method, expected_method);
+          url_loader_factory_.AddResponse(request.url.spec(),
+                                          content_to_respond);
+        }));
+  }
+
+  void OnRequestResponse(
+      bool* callback_called,
+      bool expected_success,
+      const std::string& expected_response,
+      const int http_code,
+      const std::string& body,
+      const base::flat_map<std::string, std::string>& headers) {
+    *callback_called = true;
+    bool success = http_code == 200;
+    EXPECT_EQ(expected_success, success);
+    EXPECT_EQ(expected_response, body);
+  }
+
+  void SendRequest(const std::string& server_raw_response,
+                   const std::string& expected_sanitized_response,
+                   APIRequestHelper::ResponseConversionCallback
+                       conversion_callback = base::NullCallback()) {
+    bool callback_called = false;
+    GURL network_url("http://localhost/");
+    SetInterceptor("POST", network_url, server_raw_response);
+    api_request_helper_->Request(
+        "POST", network_url, "", "application/json", false,
+        base::BindOnce(&ApiRequestHelperUnitTest::OnRequestResponse,
+                       base::Unretained(this), &callback_called,
+                       true /* success */, expected_sanitized_response),
+        {}, -1u, std::move(conversion_callback));
+    base::RunLoop().RunUntilIdle();
+    EXPECT_TRUE(callback_called);
+  }
+
+ protected:
+  std::unique_ptr<APIRequestHelper> api_request_helper_;
+
+ private:
+  base::test::TaskEnvironment task_environment_;
+  network::TestURLLoaderFactory url_loader_factory_;
+  scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory_;
+  data_decoder::test::InProcessDataDecoder in_process_data_decoder_;
+};
+
+TEST_F(ApiRequestHelperUnitTest, SanitizedRequest) {
```

Comment:
added, it returns error and empty response

Reply:
hmm, ok. that will cause problem when we apply sanitizer to ETH web3 provider because those requests often containing trailing commas and we have to add `base::JSON_ALLOW_TRAILING_COMMAS` when parsing it
cc @bbondy

Reply:
added patch to support JSON_ALLOW_TRAILING_COMMAS